tortoise/tortoise.go

```diff
@@ -636,268 +449,264 @@ func (t *turtle) getBallotBeacon(ballot *types.Ballot, logger log.Log) ([]byte,
-		globalOpinionOnBlock := calculateOpinionWithThreshold(
-			t.logger, sum, t.GlobalThreshold, weight)
-		logger.With().Debug("verifying tortoise calculated global opinion on block",
-			log.Stringer("block_voted_on", bid),
-			lid,
-			log.Stringer("global_vote_sum", sum),
-			log.Stringer("global_opinion", globalOpinionOnBlock),
-			log.Stringer("local_opinion", localOpinionOnBlock))
-
-		// At this point, we have all the data we need to make a decision on this block. There are three possible
-		// outcomes:
-		// 1. record our opinion on this block and go on evaluating the rest of the blocks in this layer to see if
-		//    we can verify the layer (if local and global consensus match, and global consensus is decided)
-		// 2. keep waiting to verify the layer (if not, and the layer is relatively recent)
-		// 3. trigger self healing (if not, and the layer is sufficiently old)
-		consensusMatches := globalOpinionOnBlock == localOpinionOnBlock
-		globalOpinionDecided := globalOpinionOnBlock != abstain
-
-		// If, for any block in this layer, the global opinion (summed block votes) disagrees with our vote (the
-		// input vector), or if the global opinion is abstain, then we do not verify this layer. This could be the
-		// result of a reorg (e.g., resolution of a network partition), or a malicious peer during sync, or
-		// disagreement about Hare success.
-		if !consensusMatches {
-			logger.With().Warning("global opinion on block differs from our vote, cannot verify layer",
-				bid,
-				log.Stringer("global_opinion", globalOpinionOnBlock),
-				log.Stringer("local_opinion", localOpinionOnBlock))
-			return nil, nil
-		}
+		previous := t.verified
+		t.verified = t.verifying.verifyLayers(logger)
 
-		// There are only two scenarios that could result in a global opinion of abstain: if everyone is still
-		// waiting for Hare to finish for a layer (i.e., it has not yet succeeded or failed), or a balancing attack.
-		// The former is temporary and will go away after `zdist' layers. And it should be true of an entire layer,
-		// not just of a single block. The latter could cause the global opinion of a single block to permanently
-		// be abstain. As long as the contextual validity of any block in a layer is unresolved, we cannot verify
-		// the layer (since the effectiveness of each transaction in the layer depends upon the contents of the
-		// entire layer and transaction ordering). Therefore we have to enter self healing in this case.
-		// TODO: abstain only for entire layer at a time, not for individual blocks (optimization)
-		//   see https://github.com/spacemeshos/go-spacemesh/issues/2674
-		if !globalOpinionDecided {
-			logger.With().Warning("global opinion on block is abstain, cannot verify layer",
-				bid,
-				log.Stringer("global_opinion", globalOpinionOnBlock),
-				log.Stringer("local_opinion", localOpinionOnBlock))
-			return nil, nil
+		if t.verified.After(previous) {
+			if err := persistContextualValidity(logger,
```

Comment:
why isn't this done inside t.verifying.verifyLayers()?

Reply:
my intention is to separate code that is required for consensus correctness from the code that loads/saves state on to the disk.

also, not related to this pr, but i think that it will be safer not to persist contextual validity in tortoise, but instead return it as in memory object to the caller, and let the caller persist at the right time.